pipeline_tools/create_envelope.py

```diff
@@ -1,277 +1,416 @@
+def add_file_reference(file_ref, file_refs_url, auth_headers, http_requests):
+    """Add a file reference to the analysis metadata in a submission envelope.
 
     Args:
-        analysis_json (dict): metadata describing the analysis
+        file_ref (dict): HCA metadata stub about the file.
+        file_refs_url (str): The url for adding file references.
+        auth_headers (dict): Dict representing headers to use for auth.
+        http_requests (http_requests.HttpRequests): The HttpRequests object to use for talking to Ingest.
 
     Returns:
-        uuid (str): A string representing the input bundle uuid
+        dict: Dict representing the JSON response to the request that adding the file reference.
+
+    Raises:
+        requests.HTTPError: For 4xx errors or 5xx errors beyond timeout.
     """
-    bundle = analysis_json['input_bundles'][0]
-    uuid = bundle
-    print('Input bundle uuid {0}'.format(uuid))
-    return uuid
+    print('Adding file: {0} to the file reference.'.format(file_ref['fileName']))
+    response = http_requests.put(file_refs_url, headers=auth_headers, json=file_ref)
+    return response.json()
 
 
-def get_output_files(analysis_json, schema_url, analysis_file_version):
-    """Get the metadata describing the outputs of the analysis
+def link_analysis_protocol_to_analysis_process(link_url, analysis_protocol_url, http_requests):
+    """Make the analysis process to be associated with the analysis_protocol to let Ingest create the links.json.
 
     Args:
-        analysis_json (dict): metadata describing the analysis
-        schema_url (str): URL for retrieving HCA metadata schemas
-        analysis_file_version (str): the schema version that file references will conform to
+        link_url (str): The url for link protocols to processes.
+        analysis_protocol_url (str): The url for creating the analysis_protocol.
+        http_requests (http_requests.HttpRequests): The HttpRequests object to use for talking to Ingest.
 
-    Returns:
-        output_refs (dict): A dict of metadata describing the output files produced by the analysis
+    Raises:
+        requests.HTTPError: For 4xx errors or 5xx errors beyond timeout.
     """
-    outputs = analysis_json['outputs']
-    output_refs = []
-
-    for out in outputs:
-        output_ref = {}
-        file_name = out['file_core']['file_name']
-        output_ref['fileName'] = file_name
-        output_ref['content'] = {
-            'describedBy': '{}/type/file/{}/analysis_file'.format(schema_url, analysis_file_version),
-            'schema_type': 'file',
-            'file_core': {
-                'file_name': file_name,
-                'file_format': out['file_core']['file_format']
-            }
-        }
-        output_refs.append(output_ref)
-    return output_refs
+    link_headers = {'content-type': 'text/uri-list'}
+    response = http_requests.put(link_url, headers=link_headers, data=analysis_protocol_url)
```

Comment:
Can you return `response.json()` at the end of this function?

Reply:
No I cannot :=( The response from Ingest for this endpoint currently returns non-JSON messages, `response.json()` will throw an error about it. Since the `response` is not used by anywhere else, I can replace it with `http_requests.put(link_url, headers=link_headers, data=analysis_protocol_url)` so it's less confusing?

Reply:
Ahh okay I wonder if that was on purpose? 🙁

Reply:
I have no idea 🤔 maybe @aaclan-ebi or @rolando-ebi can confirm?

Reply:
yup, that API is provided by Spring Data REST, and and the response for its result will be a HTTP status 204 (no content).

Reply:
@aaclan-ebi Cool, thanks for the explanation here!